### PR TITLE
non_exhaustive errors, locid, and decimal

### PR DESCRIPTION
**Pull request by Manishearth**
Split out of https://github.com/unicode-org/icu4x/pull/1789

Progress on #170, #383

### CODE REVIEW

**Comment by Manishearth** [comment]
cc @zbraniecki in case you have locid opinions

**Review by zbraniecki — DISMISSED** [review]

File: components/decimal/src/options.rs [hunk 1]
```diff
@@ -7,13 +7,24 @@
...
 pub struct FixedDecimalFormatOptions {
     /// When to render grouping separators.
     pub grouping_strategy: GroupingStrategy,
     /// When to render the sign.
     pub sign_display: SignDisplay,
 }
 
+impl FixedDecimalFormatOptions {
+    /// Construct a new [`FixedDecimalFormatOptions`] from its components
+    pub fn new(grouping_strategy: GroupingStrategy, sign_display: SignDisplay) -> Self {
```
Comment (sffc) [inline components/decimal/src/options.rs:20]: please remove, as discussed.

File: components/locid/src/extensions/transform/mod.rs [hunk 2]
```diff
@@ -79,6 +79,7 @@ use litemap::LiteMap;
 /// [`Unicode Locale Identifier`]: https://unicode.org/reports/tr35/#Unicode_locale_identifier
 #[derive(Clone, PartialEq, Eq, Debug, Default, Hash, PartialOrd, Ord)]
 #[allow(missing_docs)] // TODO(#1028) - Add missing docs.
+#[non_exhaustive]
```
Comment (sffc) [inline components/locid/src/extensions/transform/mod.rs:82]: Suggestion: this is a spec-backed data structure. I feel that it should most likely be stable. @zbraniecki ?

File: components/locid/src/extensions/unicode/mod.rs [hunk 3]
```diff
@@ -75,6 +75,7 @@ use litemap::LiteMap;
 /// ```
 #[derive(Clone, PartialEq, Eq, Debug, Default, Hash, PartialOrd, Ord)]
 #[allow(missing_docs)] // TODO(#1028) - Add missing docs.
+#[non_exhaustive]
```
Comment (sffc) [inline components/locid/src/extensions/unicode/mod.rs:78]: Suggestion: this is a spec-backed data structure. I feel that it should most likely be stable. @zbraniecki ?

**Review by sffc — APPROVED** [review]

**Comment by Manishearth** [comment]
Going to merge, if folks still have thoughts about what should and shouldn't be allow/non_exhaustive, feel free to leave comments, we still have plenty of time for 1.0. These PRs tend to bitrot easily otherwise.

**Pull request merged**